### PR TITLE
Define BeaconState Core Type and Protobuf Definition

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #781

---

# Description

Given the latest specification, there is now only a single state for the beacon chain called `BeaconState`. We will be abandoning the split between the active and crystallized states and beginning a beacon node refactor by starting from defining the new state core type. This PR encompasses a basic Golang struct to represent the BeaconState as well as protobuf definitions for the underlying types. This PR also refactors the proto directory a bit, splitting type definitions into a `types.proto` file and p2p message request/response types into a `messages.proto` file.

### CODE REVIEW

File: proto/beacon/p2p/v1/types.proto [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  bytes randao_reveal = 3;
+  bytes pow_chain_ref = 4;
+  bytes active_state_root = 5;
+  bytes crystallized_state_root = 6;
+  google.protobuf.Timestamp timestamp = 7;
+  repeated AggregatedAttestation attestations = 8;
+  repeated SpecialRecord specials = 9;
+}
+
+message CrystallizedState {
```
Comment (rauljordan) [inline proto/beacon/p2p/v1/types.proto]: We keep these in here for now until we are done with the refactor

File: proto/beacon/p2p/v1/types.proto [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+message BeaconState {
+  uint64 validator_set_change_slot = 1;
+  repeated ValidatorRecord validators = 2;
+  repeated CrosslinkRecord crosslinks = 3;
+  uint64 last_state_recalculation_slot = 4;
+  uint64 last_finalized_slot = 5;
+  uint64 last_justified_slot = 6;
+  uint64 justified_streak = 7;
+  repeated ShardAndCommitteeArray shard_and_committees_for_slots = 8;
+  repeated uint64 persistent_committees = 9;
```
Comment (terenc3t) [inline proto/beacon/p2p/v1/types.proto]: uint32

File: proto/beacon/p2p/v1/types.proto [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  bytes withdrawal_address = 3;
+  bytes randao_commitment = 4;
+  uint64 balance = 5;
+  uint64 status = 6;
+  uint64 exit_slot = 7;
+  uint64 randao_last_change = 8;
+}
+
+message ShardReassignmentRecord {
+  uint64 validator_index = 1;
```
Comment (terenc3t) [inline proto/beacon/p2p/v1/types.proto]: uint32

File: proto/beacon/p2p/v1/types.proto [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  uint64 last_justified_slot = 6;
+  uint64 justified_streak = 7;
+  repeated ShardAndCommitteeArray shard_and_committees_for_slots = 8;
+  repeated uint64 persistent_committees = 9;
+  repeated ShardReassignmentRecord persistent_committee_reassignment = 10;
+  bytes next_shuffling_seed = 11;
+  repeated uint64 deposits_penalized_in_period = 12;
+  bytes validator_delta_hash_chain = 13;
+  uint64 current_exit_seq = 14;
+  google.protobuf.Timestamp genesis_time = 15;
```
Comment (terenc3t) [inline proto/beacon/p2p/v1/types.proto]: genesis time should probably be uint64 or we have to do the translation again before ssz

**Pull request merged**